src/Migrator/MigrationWorkflow.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) InfinityFlow. All Rights Reserved.
+// Licensed under the Apache 2.0. See LICENSE file in the solution root for full license information.
+
+namespace InfinityFlow.Temporal.Migrator;
+
+using System.Reflection;
+using Abstractions;
+using Microsoft.Extensions.Logging;
+using Temporalio.Workflows;
+
+/// <summary>
+/// Migration Workflow.
+/// </summary>
+[Workflow]
+public class MigrationWorkflow
+{
+    /// <summary>
+    /// Default Task Queue Name.
+    /// </summary>
+    internal const string DefaultTaskQueueName = "migration";
+
+    private readonly Type _migrationType = typeof(IMigration);
+    private readonly List<Type> _loadedTypes = [];
+
+    /// <summary>
+    /// Run Migration.
+    /// </summary>
+    /// <para>
+    /// For <see cref="RunType.Bootstrap"/> objects expects assembly list. If objects is null it will scan all of
+    /// the assemblies available to it.
+    /// </para>
+    /// <para>
+    /// For <see cref="RunType.Migration"/> objects expects a type list. Only one is allowed, and should not be
+    /// used outside of internal calls.
+    /// </para>
+    /// <param name="runType">The run type.</param>
+    /// <param name="objects">The objects.</param>
+    /// <returns>The <see cref="Task"/>.</returns>
+    [WorkflowRun]
+    public async Task RunAsync(RunType runType = RunType.Bootstrap, string[]? objects = null)
+    {
+        switch (runType)
+        {
+            case RunType.Bootstrap:
+                await RunEntryAsync(objects ?? [], Workflow.CancellationToken);
+                break;
+            case RunType.Migration:
+                ArgumentNullException.ThrowIfNull(objects);
+                await RunMigrationAsync(objects[0], Workflow.CancellationToken);
+                break;
+            default:
+                throw new ArgumentOutOfRangeException(nameof(runType), runType, "Invalid MigrationRunner option");
+        }
+    }
+
+    private async Task RunEntryAsync(string[] assemblies, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+        var workflowId = Workflow.Info.WorkflowId;
+        var logger = Workflow.Logger;
+
+        if (assemblies.Length == 0)
+        {
+            // scan all domain
+            _loadedTypes.AddRange(AppDomain.CurrentDomain.GetAssemblies().SelectMany(assembly => assembly.GetTypes()));
```

Comment:
This is non-deterministic and should be in an activity which returns the list.

Reply:
I don't think we can achieve non-deterministic behavior here

Reply:
we absolutely can by running this in an activity